dags/sparsity_diffusion_devx/maxtext_moe_gpu_e2e.py

```diff
@@ -16,61 +16,197 @@
 
 
 import datetime
+import tempfile
+
 from airflow import models
+from airflow.decorators import task
+from airflow.hooks.subprocess import SubprocessHook
+from airflow.utils.task_group import TaskGroup
 from dags import composer_env
 from dags.common import test_owner
-from dags.common.vm_resource import XpkClusters, DockerImage
+from dags.common.vm_resource import XpkClusters, DockerImage, Project
 from dags.multipod.configs import gke_config
+from xlml.utils import gke
+
 
 # Run once a day at 11 am UTC (3 am PST)
 SCHEDULED_TIME = "0 11 * * *" if composer_env.is_prod_env() else None
 
+# Number of nodes on A3 cluster to be scaled up to
+A3_NUM_NODES = 3
+
+SCANNED_CHECKPOINT = "gs://ml-auto-solutions/output/sparsity_diffusion_devx/maxtext/chained_tests_mixtral-8x7b_nightly-2025-01-09-05-00-18//8x7b/scanned_ckpt/0/items"
+UNSCANNED_CKPT_PATH = "gs://ml-auto-solutions/output/sparsity_diffusion_devx/maxtext/chained_tests_mixtral-8x7b_nightly-2025-01-09-05-00-18//unscanned_ckpt/checkpoints/0/items"
 
-with models.DAG(
-    dag_id="maxtext_moe_gpu_e2e",
-    schedule=SCHEDULED_TIME,
-    tags=[
-        "sparsity_diffusion_devx",
-        "multipod_team",
-        "maxtext",
-        "gpu",
-        "stable",
-        "nightly",
-        "mlscale_onduty",
-    ],
-    start_date=datetime.datetime(2024, 12, 11),
-    catchup=False,
-) as dag:
-  test_name_prefix = "maxtext"
 
-  timestamp = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M")
-  train_base = (
-      "python3 MaxText/train.py MaxText/configs/base.yml model_name=mixtral-8x7b "
-      "base_output_directory=gs://runner-maxtext-logs dataset_path=gs://maxtext-dataset "
-      "steps=2 per_device_batch_size=1 hardware=gpu dataset_type=synthetic attention=cudnn_flash_te "
-      "remat_policy=full use_iota_embed=True capacity_factor=1.0 "
-      "reuse_example_batch=1 enable_checkpointing=False megablox=False "
-      "weight_dtype=bfloat16 ici_expert_parallelism=-1 ici_fsdp_parallelism=1"
+def configure_project_and_cluster(project: str, cluster_name: str, zone: str):
+  region = gke.zone_to_region(zone)
+
+  gcloud_command = (
+      f"gcloud config set project {project}",
+      "sudo chown -R airflow:airflow /home/airflow/composer_kube_config",
+      f"gcloud container clusters get-credentials {cluster_name}"
+      f"  --region {region}",
+  )
+  return gcloud_command
+
+
+def resize_a3_cluster(cluster_name: str, zone: str, num_nodes: int):
+  region = gke.zone_to_region(zone)
+  node_pool = f"{cluster_name}-np-0"
+
+  gcloud_command = (
+      f"gcloud container clusters resize {cluster_name}"
+      f"  --quiet --region {region}"
+      f"  --node-pool {node_pool}"
+      f"  --num-nodes {num_nodes}",
   )
+  return gcloud_command
+
+
+def wait_for_cluster_ready():
```

Comment:
I feel those could be refactored a little bit and put as helper function in https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/master/xlml/utils/gpu.py? including `wait_for_cluster_ready`, `scale_up_cluster` and `scale_down_cluster`, `resize_a3_cluster`, etc?